components/sceneutil/nodecallback.hpp

```diff
@@ -0,0 +1,41 @@
+#ifndef SCENEUTIL_NODECALLBACK_H
+#define SCENEUTIL_NODECALLBACK_H
+
+#include <osg/Callback>
+
+namespace osg
+{
+    class Node;
+    class NodeVisitor;
+}
+
+namespace SceneUtil
+{
+
+template <class Derived, typename NodeType=osg::Node*, typename VisitorType=osg::NodeVisitor*>
+class NodeCallback : public virtual osg::Callback
+{
+public:
+    NodeCallback(){}
+    NodeCallback(const NodeCallback& nc,const osg::CopyOp& copyop):
+            osg::Callback(nc, copyop) {}
+    META_Object(SceneUtil, NodeCallback)
+
+    virtual bool run(osg::Object* object, osg::Object* data)
+    {
+        static_cast<Derived*>(this)->operator()((NodeType)object, (VisitorType)data->asNodeVisitor());
```

Comment:
We use a traditional cast here to work with forward declared types. The ubiquitousness of callbacks in Open MW code would otherwise lead to bloated compilation units. `static_cast` benefits concerning type hierarchy verification are not necessary in this case. `traverse` would not compile for mismatching types.

Reply:
I missed that these types can be forward declarations but this file may not include their definitions, only files instantiating `NodeCallback` type do. And `osg/Node` is already included in all required files. The only problem is `osgUtil::CullVisitor`. Including it can increase compilation time if that's your concern but I don't think it will be much noticeable.

In this trade-off between compilation time and a type safety I'd choose a type safety. But I will not insist on it, maybe someone else has opinion on it.